src/effects/VST/VSTEffect.cpp

```diff
@@ -3620,47 +3627,74 @@ VSTEffectValidator::VSTEffectValidator
      mInstance(instance)
 {}
 
-// Default, do-nothing implementations of virtuals in VSTEffectWrapper
-void VSTEffectWrapper::NeedIdle()
-{   
-}
 
-void VSTEffectWrapper::UpdateDisplay()
+VSTEffectInstance& VSTEffectValidator::GetInstance()
 {
+   return mInstance;
 }
 
-VstTimeInfo* VSTEffectWrapper::GetTimeInfo()
-{
-   return nullptr;
+
+// Default, do-nothing implementations of virtuals in VSTEffectWrapper
+void VSTInstanceBase::NeedIdle()
+{   
 }
 
-void VSTEffectWrapper::SetBufferDelay(int samples)
+void VSTInstanceBase::UpdateDisplay()
 {
 }
 
-float VSTEffectWrapper::GetSampleRate()
+void VSTInstanceBase::SizeWindow(int w, int h)
 {
-   return 44100.0f;
 }
 
-int VSTEffectWrapper::GetProcessLevel()
+void VSTInstanceBase::Automate(int index, float value)
 {
-   //return mProcessLevel;
-   return 1;
+   callSetParameterB(index, value);
```

Comment:
Automate() is changed to use `callSetParameterB`, losing the loop over slaves.  That seems like a wrong change in side-effects.

You should call `callSetParameter`.   But that doesn't exist in `VSTInstanceBase`.

The fix should then be:  move Automate() to the Instance class in step 4, which is what the header file says should happen.

Also move `callSetParameter` (which includes the loop) into Instance, out of the Effect object.

But then `VSTEffect::OnSlider` won't compile.  But the solution for that will be, the Validator object will call `callSetParameter` on the Instance.  That function will ultimately be in the validator.

So make this temporary fix here in step 4:  Let the Effect retain a pointer to the last constructed Validator.  Assume there is never more than one Valdator at once and that its lifetime will include any call to `OnSlider`, which will happen only while a non-fancy, plain UI dialog is open.

Reply:
I was wrong when I said that Automate has to belong to the instance. 
It must indeed belong to the Effect, because changes in sliders are intercepted by the Effect vst callback.

That being said, yes it is still wrong that Automate() loses the loop over slaves. In commit 4.3 I fix that by having Automate use callSetParameter and that required that mInstances move to the instancebase class.